src/main/java/com/google/devtools/build/lib/analysis/RepoMappingManifestAction.java

```diff
@@ -0,0 +1,128 @@
+import net.starlark.java.eval.EvalException;
+
+/** Creates a manifest file describing the repos and mappings relevant for a runfile tree. */
+public class RepoMappingManifestAction extends AbstractFileWriteAction {
+  private static final UUID MY_UUID = UUID.fromString("458e351c-4d30-433d-b927-da6cddd4737f");
+
+  private final ImmutableList<Entry> entries;
+  private final String workspaceName;
+
+  /** An entry in the repo mapping manifest file. */
+  @AutoValue
+  public abstract static class Entry {
+    public static Entry of(
+        RepositoryName sourceRepo, String targetRepoApparentName, RepositoryName targetRepo) {
+      return new AutoValue_RepoMappingManifestAction_Entry(
+          sourceRepo, targetRepoApparentName, targetRepo);
+    }
+
+    public abstract RepositoryName sourceRepo();
+
+    public abstract String targetRepoApparentName();
+
+    public abstract RepositoryName targetRepo();
+  }
+
+  public RepoMappingManifestAction(
+      ActionOwner owner, Artifact output, List<Entry> entries, String workspaceName) {
+    super(owner, NestedSetBuilder.emptySet(Order.STABLE_ORDER), output, /*makeExecutable=*/ false);
+    this.entries =
+        ImmutableList.sortedCopyOf(
+            comparing((Entry e) -> e.sourceRepo().getName())
+                .thenComparing(Entry::targetRepoApparentName)
+                .thenComparing(e -> e.targetRepo().getName()),
+            entries);
+    this.workspaceName = workspaceName;
+  }
+
+  @Override
+  public String getMnemonic() {
+    return "RepoMappingManifest";
+  }
+
+  @Override
+  protected String getRawProgressMessage() {
+    return "writing repo mapping manifest for " + getOwner().getLabel();
+  }
+
+  @Override
+  protected void computeKey(
+      ActionKeyContext actionKeyContext,
+      @Nullable ArtifactExpander artifactExpander,
+      Fingerprint fp)
+      throws CommandLineExpansionException, EvalException, InterruptedException {
+    fp.addUUID(MY_UUID);
+    fp.addString(workspaceName);
+    for (Entry entry : entries) {
+      fp.addString(entry.sourceRepo().getName());
+      fp.addString(entry.targetRepoApparentName());
+      fp.addString(entry.targetRepo().getName());
+    }
+  }
+
+  @Override
+  public DeterministicWriter newDeterministicWriter(ActionExecutionContext ctx)
+      throws InterruptedException, ExecException {
+    return out -> {
+      Writer writer = new BufferedWriter(new OutputStreamWriter(out, ISO_8859_1));
+      for (Entry entry : entries) {
+        if (entry.targetRepoApparentName().isEmpty()) {
+          // the apparent repo name can only be empty for the main repo. we skip this line.
+          continue;
+        }
+        writer.write(entry.sourceRepo().getName());
+        writer.write(',');
+        writer.write(entry.targetRepoApparentName());
+        writer.write(',');
+        if (entry.targetRepo().isMain()) {
+          // The canonical name of the main repo is the empty string, but we use the "workspace
```

Comment:
The "we" here is a bit confusing since it could be understood as "this action", which isn't correct.